api/client/contextdialer.go

```diff
@@ -152,30 +202,35 @@ func newTLSRoutingTunnelDialer(ssh ssh.ClientConfig, keepAlivePeriod, dialTimeou
 			return nil, trace.Wrap(err)
 		}
 
-		dialer := &net.Dialer{
-			Timeout:   dialTimeout,
-			KeepAlive: keepAlivePeriod,
-		}
-		conn, err = dialer.DialContext(ctx, network, tunnelAddr)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-
 		host, _, err := webclient.ParseHostPort(tunnelAddr)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
-
-		tlsConn := tls.Client(conn, &tls.Config{
-			NextProtos:         []string{constants.ALPNSNIProtocolReverseTunnel},
-			InsecureSkipVerify: insecure,
-			ServerName:         host,
+		conn, err := DialALPN(ctx, tunnelAddr, ALPNDialerConfig{
+			DialTimeout:     params.cfg.DialTimeout,
+			KeepAlivePeriod: params.cfg.KeepAlivePeriod,
+			TLSConfig: &tls.Config{
+				NextProtos: []string{
+					ALPNProtocolWithPing(constants.ALPNSNIProtocolReverseTunnel),
+					constants.ALPNSNIProtocolReverseTunnel,
+				},
+				InsecureSkipVerify: insecure,
+				ServerName:         host,
+			},
+			ALPNConnUpgradeRequired: params.cfg.IsALPNConnUpgradeRequired(tunnelAddr, insecure),
+			GetClusterCAs: func(_ context.Context) (*x509.CertPool, error) {
+				tlsConfig, err := params.cfg.Credentials[0].TLSConfig()
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+				return tlsConfig.RootCAs, nil
+			},
```

Comment:
Could you recall why connection upgrade flow uses self signed cluster CA  instead of trusting the ALPNRouting per protocol TLS settings (forwardTLS flag) ?

Reply:
Please correct me if I am not getting the question. For reverse tunnel dial, it does not forward TLS. client was using system root CAs for the TLS handshake to verify the web tls config?